### PR TITLE
Add a cost function to take intensities into account.

### PR DESCRIPTION
**Pull request by wohe**
Adds (optional) intensities to sensor::PointCloud and
uses it in a cost function. Not yet in use.

Signed-off-by: Wolfgang Hess <whess@lyft.com>

### CODE REVIEW

File: cartographer/mapping/internal/3d/scan_matching/intensity_cost_function_3d.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#include "cartographer/sensor/point_cloud.h"
+#include "cartographer/transform/rigid_transform.h"
+#include "cartographer/transform/transform.h"
+#include "ceres/ceres.h"
+
+namespace cartographer {
+namespace mapping {
+namespace scan_matching {
+
+class IntensityCostFunction3D {
```
Comment (MichaelGrupp) [inline cartographer/mapping/internal/3d/scan_matching/intensity_cost_function_3d.h:32]: Do you want to add a small summary as we have in the other cost functions? E.g. https://github.com/cartographer-project/cartographer/blob/ee98a92845f7ded3c7e509d32f245f3860d23001/cartographer/mapping/internal/3d/scan_matching/occupied_space_cost_function_3d.h#L31

File: cartographer/mapping/internal/3d/scan_matching/intensity_cost_function_3d.cc [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#include "cartographer/mapping/internal/3d/scan_matching/intensity_cost_function_3d.h"
+
+namespace cartographer {
+namespace mapping {
+namespace scan_matching {
+
+// This method is defined here instead of the header file as it was observed
+// that defining it in the header file has a negative impact on the runtime
+// performance.
```
Comment (MichaelGrupp) [inline cartographer/mapping/internal/3d/scan_matching/intensity_cost_function_3d.cc:9]: Interesting, do you know if this is something that only applies to this function or if it also affects the other cost functions?

File: cartographer/sensor/point_cloud.h [hunk 3]
```diff
@@ -55,6 +57,9 @@ struct PointCloud {
  private:
   // For 2D points, the third entry is 0.f.
   std::vector<PointType> points_;
+  // Intensities are optional. If not empty they must have the same size as
```
[the remark this answers is not in the document]
Comment (wohe) [reply cartographer/sensor/point_cloud.h]: Done.

File: cartographer/mapping/internal/3d/scan_matching/intensity_cost_function_3d.h (hunk 1 above)
Comment (wohe) [reply cartographer/mapping/internal/3d/scan_matching/intensity_cost_function_3d.h:37]: Done.

File: cartographer/mapping/internal/3d/scan_matching/intensity_cost_function_3d.cc (hunk 2 above)
Comment (wohe) [reply cartographer/mapping/internal/3d/scan_matching/intensity_cost_function_3d.cc:9]: No idea. But a lot of attention was paid to make sure that this does not have a measurable effect on running time performance when intensities are not used.

**Review by MichaelGrupp — APPROVED** [review]

**Pull request merged**